libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -321,63 +321,59 @@ string YulUtilFunctions::roundUpFunction()
 	});
 }
 
-string YulUtilFunctions::overflowCheckedUIntAddFunction(size_t _bits)
+string YulUtilFunctions::overflowCheckedIntAddFunction(IntegerType const& _type)
 {
-	solAssert(0 < _bits && _bits <= 256 && _bits % 8 == 0, "");
-	string functionName = "checked_add_uint_" + to_string(_bits);
+	string functionName = "checked_add_" + _type.identifier();
 	return m_functionCollector->createFunction(functionName, [&]() {
 		return
 			Whiskers(R"(
 			function <functionName>(x, y) -> sum {
-				<?shortType>
-					let mask := <mask>
-					sum := add(and(x, mask), and(y, mask))
-					if and(sum, not(mask)) { revert(0, 0) }
-				<!shortType>
-					sum := add(x, y)
-					if lt(sum, x) { revert(0, 0) }
-				</shortType>
+				<?signed>
+				if and(iszero(slt(x, 0)), sgt(y, sub(<maxValue>, x))) { revert(0, 0) }
+				if and(slt(x, 0), slt(y, sub(<minValue>, x))) { revert(0, 0) }
+				<!signed>
+				if gt(x, sub(<maxValue>, y)) { revert(0, 0) }
+				</signed>
+				sum := add(x, y)
 			}
 			)")
-			("shortType", _bits < 256)
 			("functionName", functionName)
-			("mask", toCompactHexWithPrefix((u256(1) << _bits) - 1))
+			("signed", _type.isSigned())
+			("maxValue", toCompactHexWithPrefix(u256(_type.maxValue())))
+			("minValue", toCompactHexWithPrefix(u256(_type.minValue())))
 			.render();
 	});
 }
 
-string YulUtilFunctions::overflowCheckedUIntMulFunction(size_t _bits)
+string YulUtilFunctions::overflowCheckedIntMulFunction(IntegerType const& _type)
 {
-	solAssert(0 < _bits && _bits <= 256 && _bits % 8 == 0, "");
-	string functionName = "checked_mul_uint_" + to_string(_bits);
+	string functionName = "checked_mul_" + _type.identifier();
 	return m_functionCollector->createFunction(functionName, [&]() {
 		return
-			// - The current overflow check *before* the multiplication could
-			//   be replaced by the following check *after* the multiplication:
-			//   if and(iszero(iszero(x)), iszero(eq(div(product, x), y))) { revert(0, 0) }
-			// - The case the x equals 0 could be treated separately and directly return zero.
+			// Multiplication by zero could be treated separately and directly return zero.
 			Whiskers(R"(
 			function <functionName>(x, y) -> product {
-				if and(iszero(iszero(x)), lt(div(<mask>, x), y)) { revert(0, 0) }
-				<?shortType>
-					product := mulmod(x, y, <powerOfTwo>)
```

Comment:
Not sure why I used a ``mulmod`` here earlier - that's the main advantage in detecting overflows *before* the operation - if the checks pass it *won't* overflow, so we can just simply multiply here. That's the reason for removing the comment above as well, since it'd only work with ``mulmod`` here, but that's a good bit more expensive, so I'd take this as a good reason to disregard the option mentioned in the comment.